### ISSUE TITLE
mount gdrive with vfs full cache - vfs cache: failed to _ensure cache internal error: downloaders is nil

### ISSUE THREAD

**Issue by hekmon** [rclone/rclone#5984]
#### What is the problem you are having with rclone?

For several versions now, rclone mount will randomly be unable to provide read or write to file leading to OS I/O errors. I have no real clue how to replicate it other than waiting for it to happen. The only fix is to restart rclone. Frequency is random.

I managed to catch one just after a copy (logs in reverse order):
```
Feb 07 16:37:13 hostname rclone[814]: INFO  : vfs cache: cleaned: objects 2623 (was 2623) in use 87, to upload 0, uploading 0, total size 1.114Ti (was 1.114Ti)
[...]
Feb 07 16:36:06 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:06 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:06 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:06 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:05 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:05 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:05 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:05 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:05 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:05 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:05 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:05 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:05 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:05 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:05 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:05 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:05 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:36:05 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:35:58 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:35:58 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:35:45 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:35:45 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:35:45 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:35:45 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:35:45 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:35:45 hostname rclone[814]: ERROR : some/path/to/a/file: vfs cache: failed to _ensure cache internal error: downloaders is nil
Feb 07 16:35:45 hostname rclone[814]: INFO  : some/path/to/a/file: vfs cache: removed cache file as stale (remote is different)
Feb 07 16:34:22 hostname rclone[814]: INFO  : some/path/to/a/file: vfs cache: upload succeeded try #1
Feb 07 16:34:21 hostname rclone[814]: INFO  : some/path/to/a/file: Copied (new)
```

Other files are fine and can be read/write, but the affected file won't be usable until rclone restart.

#### What is your rclone version?

```
rclone v1.57.0
- os/version: ubuntu 20.04 (64 bit)
- os/kernel: 5.4.0-99-generic (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.17.2
- go/linking: static
- go/tags: none
```

#### Which OS you are using and how many bits?

Linux amd64

#### Which cloud storage system are you using? 

Google Drive

#### The command you were trying to run?

```
EnvironmentFile=/etc/rclone/%i.env
ExecStart=/usr/bin/rclone mount $SOURCE $DESTINATION --config /etc/rclone/%i.conf
```

Env file:
```
# Command
SOURCE="<redacted>"
DESTINATION="<redacted>"
WARMUPCACHE="yes"

# Global
RCLONE_CHECKERS=8
RCLONE_CHECKSUM=true
RCLONE_FAST_LIST=true
RCLONE_LOG_LEVEL=INFO
RCLONE_TRANSFERS=4
RCLONE_USE_MMAP=true

# Mount
RCLONE_ALLOW_OTHER=true
RCLONE_ATTR_TIMEOUT=8760h
RCLONE_CACHE_DIR="<redacted>"
RCLONE_DEFAULT_PERMISSIONS=true
RCLONE_DIR_CACHE_TIME=8760h
RCLONE_DIR_PERMS=0750
RCLONE_FILE_PERMS=0640
RCLONE_GID=993
RCLONE_POLL_INTERVAL=1h
RCLONE_UID=993
RCLONE_UMASK=0027
RCLONE_VFS_CACHE_MAX_AGE="2232h"
RCLONE_VFS_CACHE_MAX_SIZE="3000G"
RCLONE_VFS_CACHE_MODE="full"
RCLONE_VFS_CACHE_POLL_INTERVAL="10m"
RCLONE_VFS_READ_AHEAD="4M"
RCLONE_VFS_WRITE_BACK="1m"
RCLONE_VOLNAME="<redacted>"
RCLONE_WRITE_BACK_CACHE=true

# RC
RCLONE_RC=true
RCLONE_RC_ADDR="127.0.0.1:5573"
RCLONE_RC_USER="rclonerc"
RCLONE_RC_PASS="<redacted>"
RCLONE_RC_SERVER_WRITE_TIMEOUT=24h
```

#### A log from the command with the `-vv` flag 

This issue is already spamming hard logs on my disk by the time I see it (hundred of thousands of lines) just by being in INFO level (given this is an ERROR level spamming). As I can not force reproduce the bug and have to wait for it to happen, I would like to avoid leaving it en -vv for several days, not sure my disk can handle it (or journalctl will prune it). I *will* if you can not find anything but at start I would like to avoid it if possible.

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
Are there any other mentions of `some/path/to/a/file` in the log?

I think something like this is happening

- the file is created from scratch
- it is closed before being uploaded
- it is re-opened before being uploaded
- the file is read from
- somehow the file is marked as not completely present on disk - this is the mystery
- the error appears

If you could post the VFS meta file if it happens that would be very useful.

You can find these in your rclone cache dir, under the remote path in the `vfsMeta` directory - each is a little JSON file. There will be one for the problem file. Ideally I'd like to see its contents, and also know how long the file should be.

I don't think I'll be able to debug this without a `-vv` log. I'd prefer an unredacted log, so if you want to share one with me privately to nick@craig-wood.com then that would be best.

**Comment by hekmon** [comment]
> Are there any other mentions of `some/path/to/a/file` in the log?

Not before (file was just created) and not after (the file being unusable) except the ERROR log.

> If you could post the VFS meta file if it happens that would be very useful.

I restarted rclone to fix it so I guess the VFS meta "clean" now. Next time it occurs, I will stop rclone and copy the VFS meta somewhere else before starting it again.

> I don't think I'll be able to debug this without a `-vv` log. I'd prefer an unredacted log, so if you want to share one with me privately to [nick@craig-wood.com](mailto:nick@craig-wood.com) then that would be best.

I restarted my mount with `-vv`, w&s then. I will email you the results as soon as I got a new occurrence. Thank you!

**Comment by animosity22** [comment]
I'm able to recreate this item pretty easily.

1) Stream a file in Plex
2) Rename the file while in cache (In my case I use Sonarr or Radarr to rename) and being played
3) Error appears

I just stumbled across this as I've done a lot of renaming all my media to fix / clean up some things so I just went through those steps now to recreate it.

**Comment by hekmon** [comment]
On my end, this is a torrent client using rclone mount for the "complete" folder (incomplete torrents targets a local directory and when a file is finished, it is moved to the complete directory). So far I was not able to determinate the exact process triggering the bug.

**Comment by animosity22** [comment]
So that does make sense, it would have to be uploading from the old location, get moved to the completed and if the star align, you'd trigger it that way.

I'd imagine you can:

1) cat file
2) move it (while streaming the cat)
3) see what happens in the log 

If there is an easy way to reproduce it, there is a better chance @ncw can identify/fix it and if your exact use case is in there (even if hard to produce), it would most likely be covered as the pattern "feels" the same.

**Comment by ncw** [comment]
> I'm able to recreate this item pretty easily.
> 
> 1. Stream a file in Plex
> 2. Rename the file while in cache (In my case I use Sonarr or Radarr to rename) and being played
> 3. Error appears
> 
> I just stumbled across this as I've done a lot of renaming all my media to fix / clean up some things so I just went through those steps now to recreate it.

I used that recipe to recreate the problem

Create a 1GB file in /tmp/mount

In one terminal

    rclone -vv mount --vfs-cache-mode full --bwlimit 1M /tmp/mount/ /mnt/tmp/

In another stream the file out

    pv </mnt/tmp/1GB >/dev/null

In a third, rename the file

    mv /mnt/tmp/1GB /mnt/tmp/1GB.moved

This produces this log in the mount

```
2022/02/18 16:02:27 INFO  : 1GB: Moved (server-side) to: 1GB.moved
2022/02/18 16:02:28 DEBUG : 1GB(0xc00091e440): >_readAt: n=131072, err=<nil>
... reading continues fine until
2022/02/18 16:02:28 DEBUG : vfs cache: looking for range={Pos:27000832 Size:131072} in [{Pos:0 Size:27258880}] - present true
2022/02/18 16:02:28 DEBUG : 1GB.moved(0xc00091e440): >_readAt: n=131072, err=<nil>
2022/02/18 16:02:28 DEBUG : &{1GB.moved (rw)}: >Read: read=131072, err=<nil>
2022/02/18 16:02:28 DEBUG : &{1GB.moved (rw)}: Read: len=131072, offset=27131904
2022/02/18 16:02:28 DEBUG : 1GB.moved(0xc00091e440): _readAt: size=131072, off=27131904
2022/02/18 16:02:28 DEBUG : vfs cache: looking for range={Pos:27131904 Size:131072} in [{Pos:0 Size:27258880}] - present false
2022/02/18 16:02:28 ERROR : 1GB.moved: vfs cache: failed to _ensure cache internal error: downloaders is nil
2022/02/18 16:02:28 DEBUG : 1GB.moved: vfs cache: failed to _ensure cache internal error: downloaders is nil is not out of space
2022/02/18 16:02:28 DEBUG : 1GB.moved(0xc00091e440): >_readAt: n=0, err=internal error: downloaders is nil
2022/02/18 16:02:28 DEBUG : &{1GB.moved (rw)}: >Read: read=0, err=internal error: downloaders is nil
....
2022/02/18 16:02:28 DEBUG : vfs cache: looking for range={Pos:27254784 Size:4096} in [{Pos:0 Size:27258880}] - present true
2022/02/18 16:02:28 DEBUG : 1GB.moved(0xc00091e440): >_readAt: n=4096, err=<nil>
2022/02/18 16:02:28 DEBUG : &{1GB.moved (rw)}: >Read: read=4096, err=<nil>
2022/02/18 16:02:28 DEBUG : &{1GB.moved (rw)}: Read: len=4096, offset=27258880
2022/02/18 16:02:28 DEBUG : 1GB.moved(0xc00091e440): _readAt: size=4096, off=27258880
2022/02/18 16:02:28 DEBUG : vfs cache: looking for range={Pos:27258880 Size:4096} in [{Pos:0 Size:27258880}] - present false
2022/02/18 16:02:28 ERROR : 1GB.moved: vfs cache: failed to _ensure cache internal error: downloaders is nil
2022/02/18 16:02:28 DEBUG : 1GB.moved: vfs cache: failed to _ensure cache internal error: downloaders is nil is not out of space
2022/02/18 16:02:28 DEBUG : 1GB.moved(0xc00091e440): >_readAt: n=0, err=internal error: downloaders is nil
2022/02/18 16:02:28 DEBUG : &{1GB.moved (rw)}: >Read: read=0, err=internal error: downloaders is nil
2022/02/18 16:02:28 DEBUG : &{1GB.moved (rw)}: Read: len=4096, offset=27258880
2022/02/18 16:02:28 DEBUG : 1GB.moved(0xc00091e440): _readAt: size=4096, off=27258880
2022/02/18 16:02:28 DEBUG : vfs cache: looking for range={Pos:27258880 Size:4096} in [{Pos:0 Size:27258880}] - present false
2022/02/18 16:02:28 ERROR : 1GB.moved: vfs cache: failed to _ensure cache internal error: downloaders is nil
2022/02/18 16:02:28 DEBUG : 1GB.moved: vfs cache: failed to _ensure cache internal error: downloaders is nil is not out of space
2022/02/18 16:02:28 DEBUG : 1GB.moved(0xc00091e440): >_readAt: n=0, err=internal error: downloaders is nil
2022/02/18 16:02:28 DEBUG : &{1GB.moved (rw)}: >Read: read=0, err=internal error: downloaders is nil
... The error is returned to `pv` here and its gets an EIO
... The rename finishes at this point also as it is now unblocked by the Read having finished
2022/02/18 16:02:43 INFO  : 1GB: vfs cache: renamed in cache to "1GB.moved"
2022/02/18 16:02:43 DEBUG : 1GB.moved: Updating file with 1GB.moved 0xc000920780
2022/02/18 16:02:43 DEBUG : : Added virtual directory entry vDel: "1GB"
2022/02/18 16:02:43 DEBUG : : Added virtual directory entry vAddFile: "1GB.moved"
2022/02/18 16:02:43 DEBUG : /: >Rename: err=<nil>
2022/02/18 16:02:43 DEBUG : /: Invalidating "1GB.moved"
```

That is quite complicated...

Looking at the code I can see it is renaming the file in the cache which causes the downloaders to be set to `nil` which is the ultimate cause of the problem.

I had a go at fixing this here - can you give it a go?

[v1.58.0-beta.6000.04177102c.fix-5984-vfs-rename-open](https://beta.rclone.org/branch/fix-5984-vfs-rename-open/v1.58.0-beta.6000.04177102c.fix-5984-vfs-rename-open/) on branch [fix-5984-vfs-rename-open](https://github.com/rclone/rclone/tree/fix-5984-vfs-rename-open) (uploaded in 15-30 mins)

This needs a unit test but let's see if I'm on the right track first!

**Comment by animosity22** [comment]
I think that did what I expected.

I continued to read the file all the way through until my one iteration closed the file.

```
2022/02/18 12:54:21 DEBUG : test.mkv(0xc000364a80): >_readAt: n=131072, err=<nil>
2022/02/18 12:54:21 DEBUG : &{test.mkv (rw)}: >Read: read=131072, err=<nil>
2022/02/18 12:54:21 DEBUG : test.mkv(0xc000364a80): _readAt: size=131072, off=404094976
2022/02/18 12:54:21 DEBUG : vfs cache: looking for range={Pos:404094976 Size:131072} in [{Pos:0 Size:404291584}] - present true
2022/02/18 12:54:21 DEBUG : test.mkv(0xc000364a80): >_readAt: n=131072, err=<nil>
2022/02/18 12:54:21 DEBUG : &{test.mkv (rw)}: >Read: read=131072, err=<nil>
2022/02/18 12:54:21 DEBUG : &{test.mkv (rw)}: Read: len=65536, offset=404226048
2022/02/18 12:54:21 DEBUG : test.mkv(0xc000364a80): _readAt: size=65536, off=404226048
2022/02/18 12:54:21 DEBUG : vfs cache: looking for range={Pos:404226048 Size:65536} in [{Pos:0 Size:404291584}] - present true
2022/02/18 12:54:21 DEBUG : test.mkv(0xc000364a80): >_readAt: n=65536, err=<nil>
2022/02/18 12:54:21 DEBUG : &{test.mkv (rw)}: >Read: read=65536, err=<nil>
2022/02/18 12:54:21 DEBUG : &{test.mkv (rw)}: Read: len=4096, offset=404291584
2022/02/18 12:54:21 DEBUG : test.mkv(0xc000364a80): _readAt: size=4096, off=404291584
2022/02/18 12:54:21 DEBUG : test.mkv(0xc000364a80): >_readAt: n=0, err=EOF
2022/02/18 12:54:21 DEBUG : &{test.mkv (rw)}: >Read: read=0, err=<nil>
2022/02/18 12:54:23 DEBUG : : changeNotify: relativePath="jellyfish-400-mbps-4k-uhd-hevc-10bit.mkv", type=1
2022/02/18 12:54:23 DEBUG : : invalidating directory cache
2022/02/18 12:54:23 DEBUG : : >changeNotify:
2022/02/18 12:54:23 DEBUG : Dropbox root '': Checking for changes on remote
2022/02/18 12:54:23 DEBUG : : changeNotify: relativePath="test.mkv", type=1
2022/02/18 12:54:23 DEBUG : : >changeNotify:
2022/02/18 12:54:23 INFO  : jellyfish-400-mbps-4k-uhd-hevc-10bit.mkv: Moved (server-side) to: test.mkv
2022/02/18 12:54:23 INFO  : jellyfish-400-mbps-4k-uhd-hevc-10bit.mkv: vfs cache: renamed in cache to "test.mkv"
2022/02/18 12:54:23 DEBUG : test.mkv: Updating file with test.mkv 0xc000894900
2022/02/18 12:54:23 DEBUG : : Added virtual directory entry vDel: "jellyfish-400-mbps-4k-uhd-hevc-10bit.mkv"
2022/02/18 12:54:23 DEBUG : : Added virtual directory entry vAddFile: "test.mkv"
2022/02/18 12:54:23 DEBUG : /: >Rename: err=<nil>
2022/02/18 12:54:23 DEBUG : /: Invalidating "test.mkv"
2022/02/18 12:54:23 DEBUG : &{test.mkv (rw)}: Flush:
2022/02/18 12:54:23 DEBUG : test.mkv(0xc000364a80): RWFileHandle.Flush
2022/02/18 12:54:23 DEBUG : &{test.mkv (rw)}: >Flush: err=<nil>
2022/02/18 12:54:23 DEBUG : &{test.mkv (rw)}: Release:
2022/02/18 12:54:23 DEBUG : test.mkv(0xc000364a80): RWFileHandle.Release
2022/02/18 12:54:23 DEBUG : test.mkv(0xc000364a80): close:
2022/02/18 12:54:23 DEBUG : test.mkv: vfs cache: setting modification time to 2022-02-02 15:09:56 +0000 UTC
2022/02/18 12:54:23 DEBUG : test.mkv(0xc000364a80): >close: err=<nil>
2022/02/18 12:54:23 DEBUG : &{test.mkv (rw)}: >Release: err=<nil>
2022/02/18 12:54:23 DEBUG : /: Lookup: name="jellyfish-400-mbps-4k-uhd-hevc-10bit.mkv"
```

So I'd say that looks good to me.

**Comment by ncw** [comment]
Great - thank you for testing.

I'll try to write a unit test for the problem so it doesn't re-occur.

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.58
